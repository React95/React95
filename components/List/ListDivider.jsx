import styled from 'styled-components';

const Divider = styled.li`
  background-color: #a2a5a8;
  height: 1px;
  border-top: 1px solid #868a8e;
  border-bottom: 1px solid white;

  width: 98%;
  margin-left: 2px;
`;

export default Divider;
