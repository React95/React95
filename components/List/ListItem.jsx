import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Icon from '../Icon';
import List from './List';

import rightcaret from '../../assets/pattern/rightcaret.svg';

const Item = styled.li`
  position: relative;

  display: flex;
  align-items: center;
  margin: 0;
  padding: 2px 6px;

  i {
    margin-right: 10px;
  }

  &:hover {
    background-color: #00007f;
    color: white;
  }

  ul {
    display: none;
    position: absolute;
    top: -2px;
    left: 97%;
    color: #000;

    i {
      width: 18px;
      height: 20px;
      margin-right: 6px;
    }
  }

  svg {
    fill: white;
  }

  ${({ icon }) => !icon && 'padding-left: 26px;'};

  ${({ hasList }) => hasList
    && `
    &:after {
      position: absolute;
      content: '';
      background-color: #000;
      mask-image: url('${rightcaret}');
      mask-position: center center;
      mask-size: 5px 8px;
      mask-repeat: no-repeat;
      width: 5px;
      height: 8px;
      right: 8px;
    }

    &:hover {
      &:after {
        background-color: #FFF;
      }

      ul {
        display: block;
      }
    }
    `};
`;

const ListItem = ({
  icon, children, onClick, rest,
}) => (
  <Item
    {...rest}
    onClick={onClick}
    icon={icon}
    hasList={React.Children.map(
      children,
      children => children.type === List,
    ).some(child => child)}
  >
    {icon && <Icon name={icon} />}
    {children}
  </Item>
);
export default ListItem;
