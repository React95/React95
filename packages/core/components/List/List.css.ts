import { style, globalStyle } from '@vanilla-extract/css';
import { calc } from '@vanilla-extract/css-utils';
import { contract } from '../themes/contract.css';

import rightcaret from '../shared/imgs/rightcaret.svg';

export const listItem = style({
  position: 'relative',
  display: 'flex',
  alignItems: 'center',
  userSelect: 'none',
  margin: contract.space[0],
  paddingBlock: contract.space[2],
  paddingInline: contract.space[6],
  color: contract.colors.materialText,
  selectors: {
    '&:hover': {
      backgroundColor: contract.colors.headerBackground,
      color: contract.colors.materialTextInvert,
    },
    '&:not(:has(svg))': {
      paddingInlineStart: '26px',
    },
    '&:has(ul):after': {
      position: 'absolute',
      width: contract.space[5],
      height: contract.space[8],
      right: contract.space[8],
      content: "''",
      backgroundColor: contract.colors.materialText,
      maskImage: `url('${rightcaret}')`,
      maskPosition: 'center center',
      maskSize: `${contract.space[5]} ${contract.space[8]}`,
      maskRepeat: 'no-repeat',
    },
    '&:has(ul):hover:after': {
      backgroundColor: contract.colors.materialTextInvert,
    },
  },
});

globalStyle(`${listItem} svg`, {
  marginRight: contract.space[10],
  minWidth: contract.space[16],
});

globalStyle(`${listItem} ul`, {
  display: 'none',
  position: 'absolute',
  top: calc.negate(contract.space[2]),
  left: '97%',
  color: contract.colors.materialText,
  zIndex: contract.zIndices.taskbar,
});

globalStyle(`${listItem}:has(ul):hover > ul`, {
  display: 'block',
});

export const divider = style({
  height: contract.space[1],
  borderTopStyle: 'solid',
  borderTopWidth: contract.space[1],
  borderTopColor: contract.colors.borderDark,
  borderBottomWidth: contract.space[1],
  borderBottomStyle: 'solid',
  borderBottomColor: contract.colors.borderLightest,
  width: '98%',
  marginLeft: contract.space[2],
});

export const list = style({
  backgroundColor: contract.colors.material,
  padding: contract.space[2],
  border: 'none',
  margin: 0,
  listStyle: 'none',
  boxShadow: contract.shadows.out,
});
