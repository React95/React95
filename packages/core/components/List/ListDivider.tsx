import React, { forwardRef } from 'react';

import { Frame, FrameProps } from '../Frame/Frame';
import { divider } from './List.css';
import cn from 'classnames';

export type DividerProps = Omit<FrameProps<'li'>, 'as'>;

export const Divider = forwardRef<HTMLLIElement, DividerProps>((rest, ref) => (
  <Frame {...rest} ref={ref} className={cn(divider, rest.className)} as="li" />
));

Divider.displayName = 'List.Divider';
