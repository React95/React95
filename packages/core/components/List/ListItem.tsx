import React, { forwardRef } from 'react';
import type { ReactElement, HtmlHTMLAttributes } from 'react';
import cn from 'classnames';

import { Frame, FrameProps } from '../Frame/Frame';
import { listItem } from './List.css';

type ItemProps = Omit<FrameProps<'li'>, 'as'>;

const Item = forwardRef<HTMLLIElement, ItemProps>((rest, ref) => (
  <Frame {...rest} ref={ref} className={cn(listItem, rest.className)} as="li" />
));

export type ListItemProps = {
  icon?: ReactElement;
} & HtmlHTMLAttributes<HTMLLIElement>;

export const ListItem = forwardRef<HTMLLIElement, ListItemProps>(
  ({ icon, children = [], ...rest }, ref) => (
    <Item {...rest} ref={ref}>
      {icon}
      {children}
    </Item>
  ),
);

ListItem.displayName = 'List.Item';
