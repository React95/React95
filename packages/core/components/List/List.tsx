import React, { forwardRef } from 'react';
import cn from 'classnames';

import { ListItem } from './ListItem';
import { Divider } from './ListDivider';
import { Frame, FrameProps } from '../Frame/Frame';
import { list } from './List.css';

export type ListProps = Omit<FrameProps<'ul'>, 'as'>;

const ListRenderer = forwardRef<HTMLUListElement, ListProps>((rest, ref) => (
  <Frame {...rest} ref={ref} className={cn(list, rest.className)} as="ul" />
));

export type IListProps = typeof ListRenderer & {
  Item: typeof ListItem;
  Divider: typeof Divider;
};

export const List: IListProps = Object.assign(ListRenderer, {
  Item: ListItem,
  Divider: Divider,
});
