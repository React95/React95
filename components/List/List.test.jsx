import React from 'react';
import List from './List';
import { render } from 'react-testing-library';

describe('<List />', () => {
  describe('Snapshots', () => {
    it('should match snapshot', () => {
      const { container } = render(
        <List>
          <List.Item icon="folder_exe2">
            <List>
              <List.Item icon="microsoft_exchange">
                Microsoft Exchange
              </List.Item>
              <List.Divider />
              <List.Item icon="windows_explorer">Windows Explorer</List.Item>
            </List>
            Programs
          </List.Item>
        </List>
      );
      expect(container).toMatchSnapshot();
    });
  });
});
