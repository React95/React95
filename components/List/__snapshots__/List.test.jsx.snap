// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> Snapshots should match snapshot 1`] = `
.c2 {
  display: block;
  width: 32px;
  height: 32px;
  background-image: url('test-file-stub');
  background-repeat: no-repeat;
  background-position: center;
  background-size: contain;
}

.c1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 0;
  padding: 2px 6px;
}

.c1 i {
  margin-right: 10px;
}

.c1:hover {
  background-color: #00007f;
  color: white;
}

.c1 ul {
  display: none;
  position: absolute;
  top: -2px;
  left: 97%;
  color: #000;
}

.c1 ul i {
  width: 18px;
  height: 20px;
  margin-right: 6px;
}

.c1 svg {
  fill: white;
}

.c1:after {
  position: absolute;
  content: '';
  background-color: #000;
  -webkit-mask-image: url('test-file-stub');
  mask-image: url('test-file-stub');
  -webkit-mask-position: center center;
  mask-position: center center;
  -webkit-mask-size: 5px 8px;
  mask-size: 5px 8px;
  -webkit-mask-repeat: no-repeat;
  mask-repeat: no-repeat;
  width: 5px;
  height: 8px;
  right: 8px;
}

.c1:hover:after {
  background-color: #FFF;
}

.c1:hover ul {
  display: block;
}

.c3 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 0;
  padding: 2px 6px;
}

.c3 i {
  margin-right: 10px;
}

.c3:hover {
  background-color: #00007f;
  color: white;
}

.c3 ul {
  display: none;
  position: absolute;
  top: -2px;
  left: 97%;
  color: #000;
}

.c3 ul i {
  width: 18px;
  height: 20px;
  margin-right: 6px;
}

.c3 svg {
  fill: white;
}

.c4 {
  background-color: #a2a5a8;
  height: 1px;
  border-top: 1px solid #868a8e;
  border-bottom: 1px solid white;
  width: 98%;
  margin-left: 2px;
}

.c0 {
  background-color: #c3c7cb;
  padding: 5px 20px 6px;
  border: none;
  margin: 0;
  padding: 2px;
  list-style: none;
  box-shadow: inset 1px 1px 0px 1px #ffffff,inset 0 0 0 1px #868a8e,1px 1px 0 1px #000;
  width: 200px;
}

<div>
  <ul
    class="c0"
    width="200"
  >
    <li
      class="c1"
    >
      <i
        class="c2"
        height="32"
        name="folder_exe2"
        width="32"
      />
      <ul
        class="c0"
        width="200"
      >
        <li
          class="c3"
        >
          <i
            class="c2"
            height="32"
            name="microsoft_exchange"
            width="32"
          />
          Microsoft Exchange
        </li>
        <li
          class="c4"
        />
        <li
          class="c3"
        >
          <i
            class="c2"
            height="32"
            name="windows_explorer"
            width="32"
          />
          Windows Explorer
        </li>
      </ul>
      Programs
    </li>
  </ul>
</div>
`;
